[TV app] Store string attributes as ZCL strings

lazarkov opened the pull request:
#### Problem
TV example app was not storing properly the ZCL strings

#### Change overview
~~- Updated the `tv.zap` and enabled Test Cluster~~
- Updated TV cluster code to store ZCL strings
- Updated `test_suites.sh` script to cover newly stored ZCL strings now
- Created `TV_WakeOnLanCluster.yaml` file for Wake on LAN cluster
- Updated `TV_ApplicationBasicCluster.yaml` to cover newly stored ZCL strings

#### Testing
How was this tested? (at least one bullet point required)
- Used the `./gn_build.sh` to verify the building is successful
- Run the `./scripts/tests/test_suites.sh -a tv`

andy31415 commented on examples/tv-app/linux/include/application-basic/ApplicationBasicManager.cpp:
```diff
@@ -17,6 +17,7 @@
  */
 
 #include "ApplicationBasicManager.h"
+#include "../cluster-helper.cpp"
```
most of our other includes are using `<>`. Could we do the same here?

Can we find a different name than "helper"? Why are we including a .cpp file instead of a header file?

lazarkov replied:
Renamed now to zcl-strings.h

andy31415 commented on examples/tv-app/linux/include/cluster-helper.cpp:
```diff
@@ -0,0 +1,39 @@
+#include <app/util/af-types.h>
+
+namespace {
+template <size_t BufferLength>
```
this should be a header file instead of a c++ file.

consider naming it `zcl-strings.h` or similar. "helper" is to vague and such names end up having everything dumped into them.

lazarkov replied:
Yup. Will do

lazarkov replied:
Updated.

andy31415 reviewed (changes requested):
Request changes to use headers instead of CPP files for includes.

bzbarsky-apple commented on examples/tv-app/linux/include/zcl-strings.h:
```diff
@@ -0,0 +1,31 @@
+template <size_t BufferLength>
+uint8_t * MakeZclCharString(uint8_t (&zclString)[BufferLength], char (&cString)[BufferLength])
+{
+    static_assert(BufferLength <= 256, "Too long string to fit in ZCL_CHAR_STRING type");
```
```suggestion
    static_assert(BufferLength <= 254, "Too long string to fit in ZCL_CHAR_STRING type");
```

bzbarsky-apple replied:
I just realized this was copy/pasted from `src/app/clusters/basic/basic.cpp`.  We also have similar code in `examples/bridge-app/linux/main.cpp` (see `ToZclCharString`).

Instead of adding yet another copy of this, can we just have a single utility method?

lazarkov replied:
Created ZclString.h and addded it to `src/lib/support`

bzbarsky-apple commented on examples/tv-app/linux/include/wake-on-lan/WakeOnLanManager.cpp:
```diff
@@ -45,11 +47,13 @@ CHIP_ERROR WakeOnLanManager::Init()
-    char addressBuffer[19];
-    snprintf(addressBuffer, 19, "%c%s", 18, macAddress);
+    char cString[18];
+    strcpy(cString, macAddress);
```
This part does not make sense to me. Is the incoming thing not null-terminated?  If so, how does strcpy make sense?  And if it is, why do we need to copy it at all?  Just because the template thing doesn't realize `macAddress` is the right type due to type decay?  But then the solution is to fix the template, I'd think....

In any case, unchecked strcpy is not ok; we should always be using `strncpy` or the even safer `chip::Platform::CopyString` (aka `strlcpy`).

lazarkov replied:
I've changed the template for `ZclString/MakeZclCharString`

lazarkov commented:
> Why do you need TestCluster here?

Without it I was not able to receive stored stored server text attributes. For example macAddress, which is stored under wake-on-lan cluster.

I've removed it now and it works fine, I'm not sure where was the problem, maybe template was not built properly. Anyhow no more Test Cluster now.

bzbarsky-apple commented on src/app/clusters/basic/basic.cpp:
```diff
@@ -27,23 +27,13 @@
 #include <app/util/af.h>
 #include <app/util/attribute-storage.h>
 #include <protocols/interaction_model/Constants.h>
+#include <support/ZclString.h>
```
```suggestion
#include <lib/support/ZclString.h>
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,36 @@
+uint8_t * MakeZclCharString(uint8_t (&zclString)[BufferLength], char * cString)
+{
+    static_assert(BufferLength <= 254, "Too long string to fit in ZCL_CHAR_STRING type");
+    zclString[0] = static_cast<uint8_t>(strlen(cString));
```
So if cString is longer than 255 chars we will grab some small prefix off the front?  In particular if it's 256 chars we will end up with a zero-length string?

What should probably happen here for now, to be consistent with other places that do this stuff, is that we should use `std::min(254, strlen(cString))` as the length.

lazarkov replied:
Added tests.

lazarkov replied:
I've covered 64 char length, 254 char length and 255 char length.
In case now char is 255 length there won't be zero-length char at the end, so I set ZCL String length to 0.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,36 @@
+template <size_t BufferLength>
+uint8_t * MakeZclCharString(uint8_t (&zclString)[BufferLength], char * cString)
+{
+    static_assert(BufferLength <= 254, "Too long string to fit in ZCL_CHAR_STRING type");
```
I made a mistake earlier... the buffer we are writing to can effectively be up to 255 bytes: one byte length and 254 bytes data.  The buffer we copy from can't be bigger than 254 bytes for this to make sense.

So this assert should be:
```suggestion
    static_assert(BufferLength <= 255, "Too long string to fit in ZCL_CHAR_STRING type");
```

lazarkov replied:
Updated.

bzbarsky-apple reviewed (commented):
Are you willing to do a followup to make `examples/bridge-app/linux/main.cpp` use your new library function too?

lazarkov commented:
> Are you willing to do a followup to make `examples/bridge-app/linux/main.cpp` use your new library function too?

Updated that file as well.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,36 @@
+        len = 0;
+    }
+    zclString[0] = static_cast<uint8_t>(len);
+    memcpy(&zclString[1], cString, zclString[0]);
```
How do we know this is safe?  Nothing here guarantees `zclString` has 255 bytes of space in it (which is all we know as far as limits on data sizes go), and in fact a bunch of callsites don't have that much space...

lazarkov replied:
zclString is now a pointer so it's up to a consumer of a function to allocate the memory for the variable.

bzbarsky-apple replied:
My point is that we should not be doing that.  We should be passing in how much buffer space the consumer actually has.  Ideally by making zclString be a `MutableByteSpan`, but worst-case by having a separate argument for how much buffer we actually have.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,166 @@
+    for (size_t i = 0; i < 64; i++)
+        cString64[i] = 'A';
+
+    MakeZclCharString(zclString, cString64);
```
Worth a comment about how cString64 is null-terminated because we used calloc.

lazarkov replied:
I've removed the null terminated tests.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,166 @@
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString64 != nullptr);
+    // ZCL String has char terminated
+    NL_TEST_ASSERT(inSuite, charIsTerminated == 1);
```
What guarantees this tests true?  After the first 65 bytes, the data in zclString is random.

lazarkov replied:
Removed these tests.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,166 @@
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString254 != nullptr);
+    // ZCL String has char terminated
+    NL_TEST_ASSERT(inSuite, charIsTerminated == 1);
```
Again, I don't see why this would necessarily be 1.

lazarkov replied:
Removed these tests.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,166 @@
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString255 != nullptr);
+    // ZCL String has char terminated
+    NL_TEST_ASSERT(inSuite, charIsTerminated == 1);
```
This will test true because it will treat the 0 byte for the length as a "terminator", but I have no idea what this is even testing for....

lazarkov replied:
Removed these tests.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,36 @@
+    size_t len = strlen(cString);
+    if (len > 254)
+    {
+        len = 0;
```
I guess we could, but why would we not just clamp length to 254 instead?  Or more generally buffer size minus 1?

lazarkov replied:
The previous function with buffer size as a parameter cString[BufferSize] is now just cString *. Same as zclString, which was zclString[BufferSize] now is zclString *. 

It's due to making the ZclString function works for [this code](uint8_t * ToZclCharString(uint8_t * zclString, const char * cString, uint8_t maxLength)) here as well.

bzbarsky-apple replied:
I understand the change that was made, but what I am not sure about is why we did not change the signature to include the length of the buffer available for write, like the code linked to does....  Not passing in the buffer length here seems like it's just asking for buffer overruns.

lazarkov commented:
@bzbarsky-apple The TestZclString test cases are now renamed. 
Since ZclString is uint8_t and can receive 255 characters, I've renamed tests to: 
- less than that number `TestZclStringLessThanMaximumSize`
- equals that number `TestZclStringEqualsMaximumSize`
- bigger than that number `TestZclStringBiggerThanMaximumSize`

The goal was to verify that newly created ZclString has correct lenght with all these variations.

bzbarsky-apple reviewed (commented):
Just to reiterate: we should not be doing unsafe copies where we have no idea whether a buffer has space for what we're trying to put into it.  Consumers should tell us how much buffer we actually have to work with, ideally by passing the buffer as MutableByteSpan.

bzbarsky-apple commented:
Still waiting on fix per https://github.com/project-chip/connectedhomeip/pull/8325#discussion_r675298953

lazarkov commented:
> Still waiting on fix per [#8325 (comment)](https://github.com/project-chip/connectedhomeip/pull/8325#discussion_r675298953)

Sorry, I was busy with Test Event coding. I'll land it today.

lazarkov commented:
> > Still waiting on fix per [#8325 (comment)](https://github.com/project-chip/connectedhomeip/pull/8325#discussion_r675298953)
> 
> Sorry, I was busy with Test Event coding. I'll land it today.

Erm. "Today" was actually today.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @brief Create ZCL string from char
+ */
+static CHIP_ERROR MakeZclCharString(chip::MutableByteSpan & buffer, const char * cString)
```
Shouldn't need static inside anonymous namespace.

But also, shouldn't we just have a .cpp file for the implementation?

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,39 @@
+        len = 254;
+    }
+
+    buffer.data()[0] = static_cast<uint8_t>(len);
+    memcpy(&buffer.data()[1], cString, len);
```
We need to ensure buffer.size() is big enough for this!  Why are we not checking that?

lazarkov replied:
Now we verify the buffer size. If the size is larger than the buffer size, error will be returned.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,39 @@
+
+    buffer.data()[0] = static_cast<uint8_t>(len);
+    memcpy(&buffer.data()[1], cString, len);
+    return err;
```
Worth adding a comment that we don't need to update the length of `buffer` because consumers don't need to examine it.

lazarkov replied:
A bit tricky one. 
If the size of char is larger than 254 and I don't limit it, then we will store empty string in the buffer as it cannot fit to a maximum size.

bzbarsky-apple replied:
I think we're talking about different things. I'm talking about the fact that we are not modifying `buffer.size()` to reflect the actual amount of data we put into it because that is not something callers should be examining anyway.  We should make that clear in the API documentation for this function: if the buffer size is nonzero, the result is always safe to use as a ZCL string and the number of initialized data bytes is just stored in the first byte, with the span size being unrelated to how much data is actually in there.

bzbarsky-apple commented on examples/bridge-app/linux/main.cpp:
```diff
@@ -229,11 +218,11 @@ void HandleDeviceStatusChanged(Device * dev, Device::Changed_t itemChangedMask)
     {
-        uint8_t zclName[kUserLabelSize];
-        ToZclCharString(zclName, dev->GetName(), kUserLabelSize - 1);
+        MutableByteSpan zclName = MutableByteSpan();
```
Was this tested?  This is going to create a span with a null buffer pointer, and then the implementation of `MakeZclCharString` in this PR will crash, while one that correctly checks lengths will not copy the data correctly....  This should probably be more like this:
```suggestion
        uint8_t zclName[kUserLabelSize];
        MutableByteSpan zclNameSpan(zclName);
```
and then pass that to MakeZclCharString.

lazarkov replied:
No, it slipped as I was updating the code. 
Updated now per suggestion.

lazarkov replied:
I also relied on the CI to tell me if it's broken as I've never run/tested this app.

bzbarsky-apple commented on examples/bridge-app/linux/main.cpp:
```diff
@@ -261,7 +250,10 @@ EmberAfStatus HandleReadBridgedDeviceBasicAttribute(Device * dev, chip::Attribut
     {
-        ToZclCharString(buffer, dev->GetName(), static_cast<uint8_t>(maxReadLength - 1));
+        uint8_t bufferMemory[254];
+        MutableByteSpan zclString(bufferMemory, 254);
```
```suggestion
        MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on examples/tv-app/linux/include/application-basic/ApplicationBasicManager.cpp:
```diff
@@ -43,9 +44,13 @@ CHIP_ERROR ApplicationBasicManager::Init()
 void ApplicationBasicManager::store(chip::EndpointId endpoint, Application * application)
 {
+    uint8_t bufferMemory[64];
+    MutableByteSpan zclString(bufferMemory, 64);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on examples/tv-app/linux/include/wake-on-lan/WakeOnLanManager.cpp:
```diff
@@ -43,13 +46,14 @@ CHIP_ERROR WakeOnLanManager::Init()
-    char addressBuffer[19];
-    snprintf(addressBuffer, 19, "%c%s", 18, macAddress);
+    uint8_t bufferMemory[32];
+    MutableByteSpan zclString(bufferMemory, 32);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/app/clusters/basic/basic.cpp:
```diff
@@ -26,37 +26,29 @@
     char cString[65];
-    uint8_t zclString[65];
+    uint8_t bufferMemory[65];
+    MutableByteSpan zclString(bufferMemory, 65);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    uint8_t bufferMemory[256];
+    // Init zcl string
+    MutableByteSpan zclString = MutableByteSpan(bufferMemory);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    uint8_t bufferMemory[256];
+    // Init zcl string
+    MutableByteSpan zclString = MutableByteSpan(bufferMemory);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+    uint8_t bufferMemory[256];
+    // Init zcl string
+    MutableByteSpan zclString = MutableByteSpan(bufferMemory);
```
```suggestion
    MutableByteSpan zclString(bufferMemory);
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,142 @@
+    for (size_t i = 0; i < 255; i++)
+        cString255[i] = 'A';
+
+    MakeZclCharString(zclString, cString255);
```
Should this assert that a failure was returned (and should the previous testcases assert that success was returned)?

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    // ZCL String data does not hold empty string
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] != '\0');
+    // ZCL String is created with no errors
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
```
This assert should be the first one....

lazarkov replied:
Moved null check and error check at the assertion top.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString64 != nullptr);
+    // ZCL String length assert
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] == 64);
+    // ZCL String data is not null
+    NL_TEST_ASSERT(inSuite, zclString.data() != NULL);
+    // ZCL String data does not hold empty string
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] != '\0');
```
The order of these three asserts makes no sense.  We should assert non-null before derefencing, and why are we checking `!= 0` after we have checked `== 64` already?

lazarkov replied:
Yes it, doesn't make sense. Removed it.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    CHIP_ERROR err = MakeZclCharString(zclString, cString64);
+
+    // Memmory assert
+    NL_TEST_ASSERT(inSuite, cString64 != nullptr);
```
Uh... We already tried dereferencing it above.  This assert should presumably happen before we ever call `MakeZclCharString`?

lazarkov replied:
Moved before MakeZclCharString

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    CHIP_ERROR err = MakeZclCharString(zclString, cString64);
+
+    // Memmory assert
```
```suggestion
    // Memory assert
```

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    CHIP_ERROR err = MakeZclCharString(zclString, cString254);
+
+    // Memmory assert
```
Same comments as in`TestZclStringLessThanMaximumSize` all apply here.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    // Init zcl string
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString64 = static_cast<char *>(MemoryCalloc(1024, true));
```
```suggestion
    char * cString64 = static_cast<char *>(MemoryCalloc(1024, 1));
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    chip::Platform::MemoryFree(cString254);
+}
+
+static void TestZclStringBiggerThanMaximumSize(nlTestSuite * inSuite, void * inContext)
```
Should test sizes 255, 256, 257, just to make sure things are reasonable...  And ideally do something where we ensure we don't write off the end of the buffer, by actually having the buffer be a bit bigger, initializing it, claiming a smaller size to the callee, and then asserting that only the parts we expect to get written got written.

In particular, should test that if passed-in buffer is 0-sized we don't write things to it.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+
+    CHIP_ERROR err = MakeZclCharString(zclString, cString255);
+
+    // Memmory assert
```
Similar comments apply here as for the other tests.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    NL_TEST_ASSERT(inSuite, zclString.data() != NULL);
+    // ZCL String data does not hold empty string
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] != '\0');
+    // ZCL String is created error of inboud message too big
```
```suggestion
    // ZCL String is created error of inbound message too big
```

lazarkov replied:
Updated.

bzbarsky-apple commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,160 @@
+    // Init zcl string
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString255 = static_cast<char *>(MemoryCalloc(1024, true));
```
```suggestion
    char * cString512 = static_cast<char *>(MemoryCalloc(1024, true));
```

lazarkov commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,41 @@
+
+namespace chip {
+
+constexpr int kBufferMaximumSize = 254;
```
Updated.

lazarkov commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,28 @@
+/**
+ * @brief Create ZCL string from char
+ */
+extern CHIP_ERROR MakeZclCharString(chip::MutableByteSpan & buffer, const char * cString);
```
Updated.

lazarkov commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,28 @@
+namespace chip {
+
+/**
+ * @brief Create ZCL string from char
```
Updated.

lazarkov commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,41 @@
+
+constexpr int kBufferMaximumSize = 254;
+
+CHIP_ERROR MakeZclCharString(chip::MutableByteSpan & buffer, const char * cString)
```
Updated.

lazarkov commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,41 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    size_t len = strlen(cString);
+    if (strlen(cString) > buffer.size())
+    {
+        err = CHIP_ERROR_INBOUND_MESSAGE_TOO_BIG;
+        // set pascal string length to a maximum of kBufferMaximumSize
+        len = buffer.size() > kBufferMaximumSize ? kBufferMaximumSize : buffer.size();
+    }
```
Seems the comment is resolved with the suggested [code](https://github.com/project-chip/connectedhomeip/pull/8325#discussion_r688696147)

lazarkov commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,41 @@
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    size_t len = strlen(cString);
+    if (strlen(cString) > buffer.size())
```
Seems the comment is resolved with the suggested [code](https://github.com/project-chip/connectedhomeip/pull/8325#discussion_r688696147)

lazarkov commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,41 @@
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    size_t len = strlen(cString);
+    if (strlen(cString) > buffer.size())
+    {
+        err = CHIP_ERROR_INBOUND_MESSAGE_TOO_BIG;
+        // set pascal string length to a maximum of kBufferMaximumSize
+        len = buffer.size() > kBufferMaximumSize ? kBufferMaximumSize : buffer.size();
+    }
```
Updated per suggestion.

bzbarsky-apple commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,29 @@
+namespace chip {
+
+/**
+ * @brief Create ZCL string from char. Char can be maximum 254 characters long. In case char is longer than 254 characters,
+ *        this method store the data up to 254 characters. The rest will be deleted and error will be returned.
```
```suggestion
 * @brief Create ZCL string from char string.  If the char string is longer than 254 chars, or too long to fit into the provided buffer, this function will place as many chars as it can, up to 254, into the buffer and then return an error.
 *
 * Even if an error is returned, the buffer still holds a valid (though truncated) ZCL string, _unless_ the buffer is 0-sized, in which case it can't hold anything.
```

andy31415 commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,46 @@
+
+namespace chip {
+
+constexpr size_t kBufferMaximumSize = 254;
```
a comment sould be useful here: length prefixed data is  using the first byte for length, hence the limitation is we only support sizes of 0-255.

This comes back on why 254 though: Why can't I have 255-length strings? Instead of 255/254 could we use std limits or similar constants?

lazarkov replied:
If I'm not mistaken:
- 1 space for the null termination
- 1 space for the length
- uint8_t has space for 256 characters
--------
 Leaves us with space of 254. 

Calculating the maximum using the std numeric_limits now.

bzbarsky-apple replied:
@andy31415 @lazarkov The reason it's 254 is that in ZCL a length of 255 means something like "no string value available at all".  So if you allow a string with length 255 here, other parts of the system will treat it as either an invalid value or an empty string, depending....

bzbarsky-apple replied:
https://github.com/project-chip/connectedhomeip/pull/9486 to fix that.

andy31415 commented on src/lib/support/ZclString.h:
```diff
@@ -0,0 +1,32 @@
+
+/**
+ * @brief Create ZCL string from char string.  If the char string is longer than 254 chars, or too long to fit into the provided
+ * buffer, this function will place as many chars as it can, up to 254, into the buffer and then return an error.
```
why do we have this 'place as much as you can' behaviour? Why not error out immediately?

lazarkov replied:
Both options of keeping or discarading data looked good. It was a preference to chose to keep data and to notify user with an error.
I don't have strong option on this. So I've updated it to return nothing.

andy31415 replied:
I believe you changed the implementation, you need to update the comment as well.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,240 @@
+    // Init zcl string
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString64 = static_cast<char *>(MemoryCalloc(1024, 1));
```
Use RAII for allocation (ScopedMemoryBuffer has calloc)

Same for other tests

lazarkov replied:
Updated.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,240 @@
+    MutableByteSpan zclString(bufferMemory);
+    // Init c string
+    char * cString64 = static_cast<char *>(MemoryCalloc(1024, 1));
+    // Add random chars to c string
```
this does not seem random. How about memset(cString54, 'A', 64) ?

Same for other tests

lazarkov replied:
Yes, code changed a lot from the comment to the last line of code. Added the memset function. removed the comment.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,240 @@
+
+    // ZCL String is created error of inbound message too big
+    NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_INBOUND_MESSAGE_TOO_BIG);
+    // ZCL String data is not null
```
remove redundant comments please - this only restates the code same for the 'length assert' below. We assume that whoever is reading the code has sufficient C++ knowledge.

lazarkov replied:
Removed.

andy31415 commented on src/lib/support/ZclString.cpp:
```diff
@@ -0,0 +1,47 @@
+
+namespace chip {
+
+// maximum size of MutableByteSpan -1 (the length of pascal string)
```
the comment does not sem to really explain things: why the reference to MutableByteSpan? That one has `    size_t mDataLen;` so its size can be quite large.

I would understand a comment of:

```
// ZCL strings are stored as pascal-strings (first byte contains the length of the
// data), so the maximum string length is the maximum of uint8_t
```

however I do not understand the -1. Please explain ideally referencing some spec.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,181 @@
+ */
+
+/**
+ *    @file
```
remove this `@file` comment, it does not seem to help me understand the code. it is obvious thtat `TestZclString` contains "tests for zcl string"

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,181 @@
+{
+    uint8_t bufferMemory[1];
+    MutableByteSpan zclString(bufferMemory);
+    char * cString1 = static_cast<char *>(MemoryCalloc(1024, 1));
```
use RAII (ScopedMemoryBuffer in `ScopedBuffer.h`

andy31415 replied:
```cpp
ScopedMemoryBuffer<char> cString1;
NL_TEST_ASSERT(inSuite, cString1.Calloc(1024));

memset(cString1.Get(), 'A', 1);

// ...
```

and then don't worry about free.

lazarkov replied:
Updated.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,181 @@
+    MutableByteSpan zclString(bufferMemory);
+    char * cString64 = static_cast<char *>(MemoryCalloc(1024, 1));
+    memset(cString64, 'A', 64);
+    NL_TEST_ASSERT(inSuite, cString64 != nullptr);
```
you assert after memset - it allocation really fails, it will crash. I believe this is throughout the test.

lazarkov replied:
Updated.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,181 @@
+    chip::Platform::MemoryFree(cString254);
+}
+
+static void TestSizeZclStringBiggerThanMaximumSize_Length_255(nlTestSuite * inSuite, void * inContext)
```
This is the test I do not understand: why is a lenght of 255 unacceptable?

Generally I wonder if we  could make a test class taking in the expected err and length and parametrize the memset and such? That way we could test many lengths in loops  for example.

lazarkov replied:
Updated to have 255 acceptable as well.

andy31415 commented on src/lib/support/tests/TestZclString.cpp:
```diff
@@ -0,0 +1,181 @@
+    CHIP_ERROR err = MakeZclCharString(zclString, cString254);
+
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, zclString.data()[0] == 254);
```
Compare data as well - need to check that content was moved not just that the length was set.

lazarkov replied:
Added data verification as well.
